### PR TITLE
Tracker Parameters Schema Update

### PR DESCRIPTION
**Pull request by ianna**
- Tracker parameters schema

### CODE REVIEW

**Comment by ianna** [comment]
+1

**Comment by ianna** [comment]
@diguida - Please, use this PR to update GT.

**Comment by ianna** [comment]
@venturia - as agreed at Tracker DPG meeting, Tracker DPG approval of this schema should be expected this week. Please, comment. Thanks.

**Comment by venturia** [comment]
@ianna, I think I have no problem with these changes as long as it will continue to be easy to modify these parameters by hand and, I think, this is possible as long as it is possible to use directly the XML files. Is it possible to have them described in the README.md file of the package Geometry/TrackerCommonData  (or any other package which uses them from the xml files)? I can help to improve the description once a first version is available. Also I am not sure about the link with #7802. By the way, in the latter, has the configuration of the parameter "upgradeGeometry" been removed.

On my side I can simply say that not everything has been settled yet: the detid schema for the pixel phase 1 will change soon (as soon as we have a working workflow in 75x) but I hope this is not a problem and, actually, it will be a test of how easy it is to configure these things.

                                                                   Andrea

On Feb 18, 2015, at 17:41 , Ianna Osborne <notifications@github.com<mailto:notifications@github.com>>
 wrote:


@venturia<https://github.com/venturia> - as agreed at Tracker DPG meeting, Tracker DPG approval of this schema should be expected this week. Please, comment. Thanks.

**Comment by ianna** [comment]
@venturia - yes, new pixel has been brought up at the meeting. I'm more convinced now that changing structs to vectors in the schema would give a better flexibility. So there will be one vector per sub detector plus a vector to define other parameters. If sub detector is missing, the vector can be left empty. If it helps, I can also introduce a layer where these parameters are reinterpreted to assign specific names, but this can be done later when everything is more or less settled. Yes, "upgradeGeometry" will be removed. 

This PR is the first part of #7802 (which is being split and will be closed). This will allow @diguida produce a GT and the second part will follow without making all relval tests fail.

**Comment by venturia** [comment]
Let's discuss about that directly, maybe when I am back at CERN on March 2nd onward. In the meantime I will continue my developments on top of your, Matti's and Gaelle's changes.

                                      Andrea

On Feb 19, 2015, at 8:54 , Ianna Osborne <notifications@github.com<mailto:notifications@github.com>>
 wrote:


@venturia<https://github.com/venturia> - yes, new pixel has been brought up at the meeting. I'm more convinced now that changing structs to vectors in the schema would give a better flexibility. So there will be one vector per sub detector plus a vector to define other parameters. If sub detector is missing, the vector can be left empty. If it helps, I can also introduce a layer where these parameters are reinterpreted to assign specific names, but this can be done later when everything is more or less settled. Yes, "upgradeGeometry" will be removed.

This PR is the first part of #7802<https://github.com/cms-sw/cmssw/pull/7802> (which is being split and will be closed). This will allow @diguida<https://github.com/diguida> produce a GT and the second part will follow without making all relval tests fail.

**Comment by davidlange6** [comment]
@ianna - this would still assume that there will never be new sub detectors. Maybe you should have one struct per allowed sub detector (0-7 in the CMS DetId schema)?

**Comment by ianna** [comment]
@davidlange6 - yes, this is a good idea. We can have 9 vectors defined. We can always switch to a new record quite easily.

**Comment by venturia** [comment]
The subdetector will "never" be more than six at the same time.

                      Andrea

On Feb 19, 2015, at 9:02 , Ianna Osborne <notifications@github.com<mailto:notifications@github.com>>
 wrote:


@davidlange6<https://github.com/davidlange6> - yes, this is a good idea. We can have 9 vectors defined. We can always switch to a new record quite easily.

**Comment by ianna** [comment]
@venturia - so, there will be only six vectors filled in (plus the odd one) and the rest of them empty. This will also be an indication of Tracker configuration.

**Comment by ianna** [comment]
@davidlange6 - this is a much more flexible implementation. A vector of Items stored where every item has GeometricDet::GDEnumType id and a vector of parameters.

@diguida, @ggovi, @apfeiffer1 - please, comment

To write a tag:
```
cmsRun CondTools/Geometry/test/trackergeometrywriter.py
```
to test parameters stored in it:
```
cmsRun Geometry/TrackerGeometryBuilder/test/trackerParameters_cfg.py
```

**Comment by venturia** [comment]
Yana,

 I am not sure I understand why GeometricDet::GDEnumType has to appear in the configuration. The assignment of a GeometricDet::GDEnumType to a subdetector is (was?) done by the code using the name of the subdetector (from the XML file).

                       Andrea

On Feb 19, 2015, at 12:26 , Ianna Osborne <notifications@github.com<mailto:notifications@github.com>>
 wrote:


@davidlange6<https://github.com/davidlange6> - this is a much more flexible implementation. A vector of Items stored where every item has GeometricDet::GDEnumType id and a vector of parameters.

@diguida<https://github.com/diguida>, @ggovi<https://github.com/ggovi>, @apfeiffer1<https://github.com/apfeiffer1> - please, comment

To write a tag:

cmsRun CondTools/Geometry/test/trackergeometrywriter.py


to test parameters stored in it:

cmsRun Geometry/TrackerGeometryBuilder/test/trackerParameters_cfg.py

**Comment by venturia** [comment]
OK now I am looking more carefully and I see something I don't like: GeometricDet must not be used to index the six subdetectors. GeometricDet is our only chance to have more than six subdetectors described in a single release. To count the six subdetectors you have to use the enumerators used by the Strip and Pixel DetId to index the subdet.
I am sorry but it will take some time to me to look at these changes.

              Andrea

On Feb 19, 2015, at 12:26 , Ianna Osborne <notifications@github.com<mailto:notifications@github.com>>
 wrote:


@davidlange6<https://github.com/davidlange6> - this is a much more flexible implementation. A vector of Items stored where every item has GeometricDet::GDEnumType id and a vector of parameters.

@diguida<https://github.com/diguida>, @ggovi<https://github.com/ggovi>, @apfeiffer1<https://github.com/apfeiffer1> - please, comment

To write a tag:

cmsRun CondTools/Geometry/test/trackergeometrywriter.py


to test parameters stored in it:

cmsRun Geometry/TrackerGeometryBuilder/test/trackerParameters_cfg.py

**Comment by apfeiffer1** [comment]
>
> \@davidlange6 <https://github.com/davidlange6> - this is a much more
> flexible implementation. A vector of Items stored where every item has
> GeometricDet::GDEnumType id and a vector of parameters.
>
> @diguida <https://github.com/diguida>, @ggovi <https://github.com/ggovi>,
> @apfeiffer1 <https://github.com/apfeiffer1> - please, comment
>

as stated in a private mail, this kind of unstructured "blob" constructs
can cause maintenance issues in the future.

Other than that - and since I don't have to maintain the code that uses
this - I'm fine with it ...

       cheers, andreas

File: Geometry/TrackerGeometryBuilder/src/TrackerParametersFromDD.cc [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+#include "DetectorDescription/Core/interface/DDCompactView.h"
+#include "DetectorDescription/Core/interface/DDVectorGetter.h"
+#include "DetectorDescription/Base/interface/DDutils.h"
+
+bool
+TrackerParametersFromDD::build( const DDCompactView* cvp,
+				PTrackerParameters& ptp)
+{
+  std::vector<int> pxbPars = dbl_to_int( DDVectorGetter::get( "pxbPars" ));
+  putOne( GeometricDet::PixelBarrel, pxbPars, ptp );
```
Comment (venturia) [inline Geometry/TrackerGeometryBuilder/src/TrackerParametersFromDD.cc]: Better not to use GeometricDet::… enumerators here. Either use 1,2,…,6 or PixelSubdetector::PixelBarrel, PixelSubdetector::PixelEndcap, StripSubdetector::TIB, StripSubdetector::TID, StripSubdetector::TOB, StripSubdetector::TEC

Comment (ianna) [inline Geometry/TrackerGeometryBuilder/src/TrackerParametersFromDD.cc]: ok, integers then.

**Comment by ianna** [comment]
@venturia - we can discuss it when you are at CERN.

**Comment by ianna** [comment]
To produce tracker parameters tag for run 1, 2, and 3:
```
cmsRun CondTools/Geometry/test/trackergeometrywriter.py
cmsRun CondTools/Geometry/test/trackergeometrywriterRun2.py
cmsRun CondTools/Geometry/test/trackergeometrywriterRun3.py
```
NOTE: PR#7779 is needed for run 3.

To analyse tags:
```
cmsRun Geometry/TrackerGeometryBuilder/test/trackerParametersRun3FromDDD_cfg.py
cmsRun Geometry/TrackerGeometryBuilder/test/trackerParametersRun2FromDDD_cfg.py
cmsRun Geometry/TrackerGeometryBuilder/test/trackerParametersFromDDD_cfg.py
```

**Comment by ianna** [comment]
+1

**Comment by ianna** [comment]
@cerminar, @ggovi,  @diguida, @mmusich, @apfeiffer1 - the schema is good to go. Does it make any difference which CMSSW version should be used for producing tags? Please, expect a GT update request shortly.

**Comment by apfeiffer1** [comment]
Hi Yana,

> @cerminar <https://github.com/cerminar>, @ggovi <https://github.com/ggovi>,
> @diguida <https://github.com/diguida>, @mmusich
> <https://github.com/mmusich>, @apfeiffer1 <https://github.com/apfeiffer1>
> - the schema is good to go. Does it make any difference which CMSSW version
> should be used for producing tags? Please, expect a GT update request
> shortly.
>
>
please do NOT use any of CMSSW_7_4_0_preN where N<7 - these will break
older releases reading the tags. pre7 is OK as is one of the more recent
IBs.

Thanks,
       cheers, andreas

PS: just to remind you that I don't have approval status any more ... ;)

**Comment by ianna** [comment]
@ggovi and @apfeiffer1 - I've produced the payloads with CMSSW_7_5_X_2015-02-25-1400, but they've failed to upload to Prod DB with complaint that the tag does not match the schema. This is the first upload to Prod DB. Are there any steps needed to update DropBox? Thanks.

**Comment by ggovi** [comment]
Hi Yana,

we can't update the dropbox with an IB. We need a pre release with your changes.  In this first phase, you should test the new tag with an sqlite and set it in the as_prefer in your job cfg.

Cheers
Giacomo

**Comment by ianna** [comment]
@davidlange6 - Could you, please, make sure this PR and corresponding #7919 are in the next pre-releases? Thanks.

**Comment by davidlange6** [comment]
@ianna,

once alca and db sign them, they will be in the IB (as usual - starting with 75x)

**Comment by ianna** [comment]
@cerminar, @ggovi, @diguida, @mmusich - ping

**Comment by ianna** [comment]
@cerminar, @ggovi, @diguida, @mmusich - please, let me know why this is not signed yet. Is there a problem with it? Thanks.

**Comment by ianna** [comment]
@cerminar, @ggovi, @diguida, @mmusich - I hope, you'll find time to test and sign it before the deadline for pre1 (tomorrow) as it is needed for you to update DropBox. Thanks.

**Comment by diguida** [comment]
+1
@ianna do you have a backport in 74X?

**Comment by ianna** [comment]
@diguida - yes, it's PR #7919

**Comment by ggovi** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**